### PR TITLE
Permit overriding find_program from the cross file.

### PR DESCRIPTION
**Pull request by jpakkane**
Part of #2005.

### CODE REVIEW

File: run_unittests.py [hunk 1]
```diff
@@ -1907,6 +1907,27 @@ def test_pch_with_address_sanitizer(self):
...
+[properties]
+
+[host_machine]
+system = 'linux'
+cpu_family = 'arm'
+cpu = 'armv7' # Not sure if correct.
+endian = 'little'
+''' % os.path.join(testdir, 'some_cross_tool.py'))
+        crossfile.flush()
+        self.init(testdir, ['--cross-file='+crossfile.name])
```
Comment (ignatenkobrain) [inline run_unittests.py]: __[flake8]__
* [E226] missing whitespace around arithmetic operator

File: mesonbuild/interpreter.py [hunk 2]
```diff
@@ -1974,8 +1979,25 @@ def func_find_program(self, node, args, kwargs):
...
+        if not args:
+            raise InterpreterException('No program name specified.')
+        required = kwargs.get('required', True)
+        if not isinstance(required, bool):
+            raise InvalidArguments('"required" argument must be a boolean.')
+        progobj = None
+        if self.build.environment.is_cross_build():
+            use_native = kwargs.get('native', False)
+            if not use_native:
+                 progobj = self.program_from_cross_file(args)
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[flake8]__
* [E111] indentation is not a multiple of four

File: mesonbuild/interpreter.py [hunk 3]
```diff
@@ -1974,8 +1985,25 @@ def func_find_program(self, node, args, kwargs):
...
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[flake8]__
* [E111] indentation is not a multiple of four

File: docs/markdown/Reference-manual.md [hunk 4]
```diff
@@ -460,6 +460,10 @@ first one that is found is returned. Meson versions earlier than
 
 Keyword arguments are the following:
 
+- `native` tells whether to find the executable from the system or use
+  the one defined in the cross file, defaults to `false` and is
+  ignored in non-cross builds
+
```
Comment (nirbheek) [inline docs/markdown/Reference-manual.md]: I would rephrase this to explicitly state what the action will be when this is set to `true`. It's not clear to me what it's supposed to do as-is.

Does `native: true` mean it will only look for executables that can be run on the build machine? `false` (the default) will cause it to also use the one defined in the cross file? Which one takes priority if the program is on the build machine and also in the cross file?

File: mesonbuild/interpreter.py [hunk 5]
```diff
@@ -1974,8 +1985,25 @@ def func_find_program(self, node, args, kwargs):
...
+    @permittedKwargs(permitted_kwargs['find_program'])
+    def func_find_program(self, node, args, kwargs):
+        if not args:
+            raise InterpreterException('No program name specified.')
+        required = kwargs.get('required', True)
+        if not isinstance(required, bool):
+            raise InvalidArguments('"required" argument must be a boolean.')
+        progobj = None
+        if self.build.environment.is_cross_build():
+            use_native = kwargs.get('native', False)
```
Comment (nirbheek) [inline mesonbuild/interpreter.py]: I think this should be subject to the same booleanness check as `required` above, just for strictness.

File: test cases/common/159 reserved targets/phony/meson.build [hunk 6]
```diff
@@ -1 +1 @@
-executable('test-phony', '../test.c')
+executable('test-PHONY', '../test.c')
```
Comment (nirbheek) [inline test cases/common/159 reserved targets/phony/meson.build]: Is this change obsolete now?

File: test cases/unit/12 cross prog/some_cross_tool.py [hunk 7]
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/env python
```
Comment (nirbheek) [inline test cases/unit/12 cross prog/some_cross_tool.py]: Let's make this python3 and force the use of python3 everywhere? Same for the other tool.

**Comment by nirbheek** [comment]
Overall, I really like this, and it's a good step towards providing a way to generically override `find_program()` with executables built by the project itself.

File: mesonbuild/interpreter.py [hunk 8]
```diff
@@ -1975,8 +1980,27 @@ def func_find_program(self, node, args, kwargs):
...
+            if not isinstance(use_native, bool):
+                raise InvalidArguments('Argument to "native" must be a boolean.')
+            if not use_native:
+                 progobj = self.program_from_cross_file(args)
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[flake8]__
* [E111] indentation is not a multiple of four

**Comment by jpakkane** [comment]
Updated.

**Comment by nirbheek** [comment]
Test failure seems to not be spurious

**Comment by jpakkane** [comment]
Of course the tests pass locally. Grrr. What could be causing this now?

**Comment by jpakkane** [comment]
The OSX failure _is_ spurious.
